src/main/java/hudson/plugins/ec2/EC2CleanupOrphanedNodes.java

```diff
@@ -0,0 +1,274 @@
+     */
+    private Set<String> updateLocalInstancesTag(
+            Ec2Client connection, Set<String> localInstanceIds, Set<String> remoteInstanceIds, EC2Cloud cloud) {
+        if (localInstanceIds.isEmpty()) {
+            LOGGER.fine(() -> "No local EC2 agents found, skipping tag update.");
+            return Set.of();
+        }
+
+        Set<String> instanceIdsToUpdate = Sets.intersection(remoteInstanceIds, localInstanceIds);
+
+        if (instanceIdsToUpdate.isEmpty()) {
+            LOGGER.fine(() -> "No local EC2 agents found in remote instances, skipping tag update.");
+            return Set.of();
+        }
+
+        LOGGER.fine(() -> "Updating tags for " + instanceIdsToUpdate.size() + " instances");
+        createOrUpdateExpiryTagInBulk(connection, cloud, instanceIdsToUpdate);
+        return instanceIdsToUpdate;
+    }
+
+    private void createOrUpdateExpiryTagInBulk(Ec2Client connection, EC2Cloud cloud, Set<String> instancesToTag) {
+
+        String nodeExpiresAtTagValue = OffsetDateTime.now(ZoneOffset.UTC)
+                .plus(RECURRENCE_PERIOD * LOST_MULTIPLIER, ChronoUnit.MILLIS)
+                .toString();
+
+        // Split instancesToTag into batches to avoid exceeding AWS limits
+        List<List<String>> batches = Lists.partition(new ArrayList<>(instancesToTag), 500);
+        LOGGER.fine(() ->
+                "Creating or updating tags in batches of " + batches.size() + " for cloud: " + cloud.getDisplayName());
+        for (List<String> batch : batches) {
+            try {
+                connection.createTags(builder -> builder.resources(batch)
+                        .tags(Tag.builder()
+                                .key(NODE_EXPIRES_AT_TAG_NAME)
+                                .value(nodeExpiresAtTagValue)
+                                .build())
+                        .build());
+                LOGGER.finer(() -> "Created or Updated tag for instances " + batch + " to " + nodeExpiresAtTagValue
+                        + " in cloud: " + cloud.getDisplayName());
+            } catch (SdkException e) {
+                LOGGER.log(Level.WARNING, "Error updating tags for instances " + batch, e);
+            }
+        }
+    }
+
+    private boolean isOrphaned(Instance remote) {
+        String nodeExpiresAt;
+        if (remote.tags() != null) {
+            nodeExpiresAt = remote.tags().stream()
+                    .filter(tag -> NODE_EXPIRES_AT_TAG_NAME.equals(tag.key()))
+                    .map(Tag::value)
+                    .findFirst()
+                    .orElse(null);
+        } else {
+            nodeExpiresAt = null;
+        }
+
+        if (nodeExpiresAt == null) {
+            LOGGER.fine(() -> "Instance " + remote.instanceId() + " does not have the tag " + NODE_EXPIRES_AT_TAG_NAME);
+            return false;
+        }
+        String currnetTime = OffsetDateTime.now(ZoneOffset.UTC).toString();
+        boolean isOrphan = nodeExpiresAt.compareTo(currnetTime) < 0;
```

Comment:
ok I guess, the String#compareTo works because both the `nodeExpiresAt` and `currentTime` are in the same format - but I think it's prone to bugs if the format of one of these values is to change (ex: if the string length itself is different the comparison would go wrong)

Reply:
ok